pkg/apis/kafka/v1alpha1/reset_offset_types.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+const (
+	OffsetEarliest = "earliest"
+	OffsetLatest   = "latest"
+)
+
+// +genclient
+// +genreconciler
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResetOffset is a resource representing a "command" for re-positioning the
+// offsets of a specific Kafka resource (Subscription, Trigger, etc.)
+type ResetOffset struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the ResetOffset.
+	Spec ResetOffsetSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the ResetOffset.
+	// This data may be out of date.
+	// +optional
+	Status ResetOffsetStatus `json:"status,omitempty"`
+}
+
+var (
+	// Check that this resource can be validated and defaulted.
+	_ apis.Validatable = (*ResetOffset)(nil)
+	_ apis.Defaultable = (*ResetOffset)(nil)
+
+	_ runtime.Object = (*ResetOffset)(nil)
+
+	// Check that we can create OwnerReferences to an this resource.
+	_ kmeta.OwnerRefable = (*ResetOffset)(nil)
+
+	// Check that the type conforms to the duck Knative Resource shape.
+	_ duckv1.KRShaped = (*ResetOffset)(nil)
+)
+
+// ResetOffsetSpec defines the specification for a ResetOffset.
+type ResetOffsetSpec struct {
+
+	// Offset is a String representing the desired offset position to which all partitions
+	// will be reset.  Supported values include "earliest", "latest", or a valid date / time
+	// string in the time.RFC3339 format. The "earliest" and "latest" values indicate the
+	// beginning and end, respectively, of the persistence window of the Topic.  There is no
+	// default value, and invalid values will result in the ResetOffset operation being
+	// rejected as failed.
+	Offset string `json:"offset"`
```

Comment:
I think the name here is straightforward, and fits the use case, so I like it.

However, for the sake of future-proof, I cannot think how to name an eventual future "numeric" offset change... How would you do that @travis-minke-sap?

Did you thought to something like:

```
offset:
  time: [earliest|latest|time string]
```

```
offset:
  value: 10
```

Reply:
I was just thinking we would attempt to parse the string value as an int64 and see if it was successful (assuming the other checks failed to parse).   I'm fine with making it more explicit though (via the field typing/separation as you suggest).  This will just mean some additional "overhead" of another json struct, etc. which is fine.  I'll make the change.

Reply:
Also - in your example it wouldn't be a simple "value" but would instead have to be an array/map of values for each Partition, which is an even better reason to do as you suggested ; )

Reply:
> This will just mean some additional "overhead" of another json struct, etc. which is fine. I'll make the change.

For the sake of forward compatibility with future additions to the API, I think that's a reasonable overhead. Now just add the `time` field (or just choose another name you prefer), then we'll reiterate on it when we want to further expand the PR.